integration/operator_test.go

```diff
@@ -0,0 +1,232 @@
+	if err == nil {
+		return nil
+	}
+
+	namespace := &v1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "giantswarm",
+		},
+	}
+	_, err = cs.CoreV1().Namespaces().Create(namespace)
+	if err != nil {
+		return err
+	}
+
+	return waitFor(activeNamespaceFunc(cs, "giantswarm"))
+}
+
+func installVault(cs kubernetes.Interface) error {
+	if err := runCmd("helm registry install quay.io/giantswarm/vaultlab-chart:stable -- --set vaultToken=${VAULT_TOKEN} -n vault"); err != nil {
+		return err
+	}
+
+	return waitFor(runningPodFunc(cs, "default", "app=vault"))
+}
+
+func installCertOperator(cs kubernetes.Interface) error {
+	installationValues := `commonDomain: ${COMMON_DOMAIN}
+clusterName: ${CLUSTER_NAME}
+Installation:
+  V1:
+    Auth:
+      Vault:
+        Address: http://vault.default.svc.cluster.local:8200
+        CA:
+          TTL: 1440h
+    Guest:
+      Kubernetes:
+        API:
+          EndpointBase: ${COMMON_DOMAIN}
+    Secret:
+      CertOperator:
+        SecretYaml: |
+          service:
+            vault:
+              config:
+                token: ${VAULT_TOKEN}
+      Registry:
+        PullSecret:
+          DockerConfigJSON: "{\"auths\":{\"quay.io\":{\"auth\":\"$REGISTRY_PULL_SECRET\"}}}"
+`
+	installationValuesEnv := os.ExpandEnv(installationValues)
+	if err := ioutil.WriteFile(certOperatorValuesFile, []byte(installationValuesEnv), os.ModePerm); err != nil {
+		return err
+	}
+	if err := runCmd("helm registry install quay.io/giantswarm/cert-operator-chart@1.0.0-${CIRCLE_SHA1} -- -n cert-operator --values " + certOperatorValuesFile); err != nil {
+		return err
+	}
+
+	return waitFor(tprFunc(cs, certificatetpr.Name))
+}
+
+func runCmd(cmdStr string) error {
+	cmdEnv := os.ExpandEnv(cmdStr)
+	fields := strings.Fields(cmdEnv)
+	cmd := exec.Command(fields[0], fields[1:]...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stdout
+
+	return cmd.Run()
+}
+
+func waitFor(f func() error) error {
+	timeout := time.After(defaultDeadline * time.Second)
+	ticker := backoff.NewTicker(backoff.NewExponentialBackOff())
+
+	for {
+		select {
+		case <-timeout:
+			ticker.Stop()
+			return fmt.Errorf("wait timeout")
+		case <-ticker.C:
+			if err := f(); err == nil {
+				return nil
+			}
+		}
+	}
+}
+
+func runningPodFunc(cs kubernetes.Interface, namespace, labelSelector string) func() error {
+	return func() error {
+		pods, err := cs.CoreV1().Pods(namespace).List(metav1.ListOptions{
+			LabelSelector: labelSelector,
+		})
+		if err != nil {
+			return err
+		}
+		if len(pods.Items) != 1 {
+			return fmt.Errorf("unexpected number of pods")
+		}
+		pod := pods.Items[0]
+		phase := pod.Status.Phase
+		if phase != v1.PodRunning {
+			return fmt.Errorf("unexpected pod status phase " + string(phase))
+		}
+		return nil
+	}
+}
+
+func activeNamespaceFunc(cs kubernetes.Interface, name string) func() error {
+	return func() error {
+		ns, err := cs.CoreV1().Namespaces().Get(name, metav1.GetOptions{})
+
+		if err != nil {
+			return err
```

Comment:
I would like to use error masking. I want to see where errors are coming from when shit hits the fan.

Reply:
Done